### ISSUE TITLE
Wrong block is being approved on ChangeView

### ISSUE THREAD

**Issue by roman-khimov** [nspcc-dev/dbft#18]
nspcc-dev/neo-go#634 is actually a dbft library problem.

**Comment by roman-khimov** [comment]
Still there as of b135b975faf370add18a71b5dbe1abd33abf0bc4 (#22)
```
2020-01-31T17:57:27.643Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:552     reset timer     {"h": 6016, "v": 0, "delay": "5s"}
2020-01-31T17:57:27.644Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/check.go:29     check preparations      {"hasReq": true, "count": 3, "M": 3}
2020-01-31T17:57:27.654Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/send.go:9       broadcasting message    {"type": "Commit", "height": 6016, "view": 0}
2020-01-31T17:57:27.655Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:552     reset timer     {"h": 6016, "v": 0, "delay": "5s"}
2020-01-31T17:57:27.657Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/check.go:64     approving block {"height": 6016, "hash": "08791652f8bfe5822ecbb74f147802f75b8ef7b5fb1b7f9cc5a3ab2d71965a2e", "tx_count": 8760, "merkle": "152eacbf199650de0ca8623c226d8d329b822fd12f532d0bf14ea55e6f92b14e", "prev": "34c4bc3c1ee70e0a88e71d5f76bc8a57f28bcad201b0188bc1c7e6756e1d865e"}
2020-01-31T17:57:27.671Z        WARN    consensus/consensus.go:342      error on add block      {"error": "block 2e5a96712daba3c59c7f1bfbb5f78e5bf70278144fb7cb2e82e5bff852167908 is invalid: signature check failed"}
github.com/CityOfZion/neo-go/pkg/consensus.(*service).processBlock
        github.com/CityOfZion/neo-go@/pkg/consensus/consensus.go:342
github.com/nspcc-dev/dbft.(*DBFT).checkCommit
        github.com/nspcc-dev/dbft@v0.0.0-20200131143342-b135b975faf3/check.go:71
github.com/nspcc-dev/dbft.(*DBFT).checkPrepare
        github.com/nspcc-dev/dbft@v0.0.0-20200131143342-b135b975faf3/check.go:36
github.com/nspcc-dev/dbft.(*DBFT).sendPrepareRequest
        github.com/nspcc-dev/dbft@v0.0.0-20200131143342-b135b975faf3/send.go:41
github.com/nspcc-dev/dbft.(*DBFT).onRecoveryMessage
        github.com/nspcc-dev/dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:532
github.com/nspcc-dev/dbft.(*DBFT).OnReceive
        github.com/nspcc-dev/dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:245
github.com/CityOfZion/neo-go/pkg/consensus.(*service).eventLoop
        github.com/CityOfZion/neo-go@/pkg/consensus/consensus.go:208
```

**Comment by roman-khimov** [comment]
Previous block:
```
2020-01-31T17:56:55.909Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:104     initialize      {"height": 6015, "view": 0, "index": 0, "role": "Backup"}
2020-01-31T17:56:55.909Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:552     reset timer     {"h": 6015, "v": 0, "delay": "10s"}
2020-01-31T17:56:55.909Z        DEBUG   consensus/consensus.go:172      timer fired     {"height": 6014, "view": 0}
2020-01-31T17:56:55.909Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:167     timeout: ignore old timer       {"height": 6014, "view": 0}
2020-01-31T17:56:56.035Z        INFO    blockchain persist completed    {"persistedBlocks": 1, "persistedKeys": 18716, "headerHeight": 6014, "blockHeight": 6014, "took": "54.052678ms"}
2020-01-31T17:56:56.473Z        DEBUG   consensus/consensus.go:207      received message        {"from": 3, "type": "PrepareResponse"}
2020-01-31T17:56:56.473Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "PrepareResponse", "from": 3, "height": 6014, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:56:56.473Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:212     ignoring old height     {"height": 6014}
2020-01-31T17:56:56.473Z        DEBUG   consensus/consensus.go:207      received message        {"from": 3, "type": "Commit"}
2020-01-31T17:56:56.473Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "Commit", "from": 3, "height": 6014, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:56:56.473Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:212     ignoring old height     {"height": 6014}
2020-01-31T17:56:56.479Z        DEBUG   consensus/consensus.go:254      payload is already in cache     {"hash": "8cb8b5382534b98c76fc345608f3e24a73ba4b7bcb0d4bbd8510fc7bcd0756b4", "type": "PrepareResponse"}
2020-01-31T17:56:56.480Z        DEBUG   consensus/consensus.go:254      payload is already in cache     {"hash": "7d4f03e1645064e0275f0f82a2e48f977d7340327c05421219730735bf9952ba", "type": "Commit"}
2020-01-31T17:57:04.102Z        DEBUG   consensus/consensus.go:207      received message        {"from": 1, "type": "RecoveryRequest"}
2020-01-31T17:57:04.102Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "RecoveryRequest", "from": 1, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:04.561Z        DEBUG   consensus/consensus.go:207      received message        {"from": 2, "type": "RecoveryMessage", "#preparation": 0, "#commit": 0, "#changeview": 0, "#request": false, "#hash": false}
2020-01-31T17:57:04.561Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "RecoveryMessage", "from": 2, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:04.561Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:492     recovery message received       {"dump": {"Type":65,"ViewNumber":0,"Witness":{"invocation":"40c1957211d7324e676a706bab698fc90bcef95bba8eeed22f293cf751801607895638319b03d717b27377c0e8ecaef562d017bd06d1e2a209208449630919dea1","verification":"2103d90c07df63e690ce77912e10ab51acc944b66860237b608c4f8f8309e71ee699ac"}}}
2020-01-31T17:57:04.561Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:506     recovering finished cv=0/4 preq=0/0 presp=0/4 co=0/4
2020-01-31T17:57:05.909Z        DEBUG   consensus/consensus.go:172      timer fired     {"height": 6015, "view": 0}
2020-01-31T17:57:05.909Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:174     timeout {"height": 6015, "view": 0}
2020-01-31T17:57:05.909Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:552     reset timer     {"h": 6015, "v": 0, "delay": "20s"}
2020-01-31T17:57:05.909Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/send.go:73      request change view     {"view": 0, "height": 6015, "new_view": 1, "nc": 0, "nf": 1}
2020-01-31T17:57:05.909Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/send.go:9       broadcasting message    {"type": "ChangeView", "height": 6015, "view": 0}
2020-01-31T17:57:06.001Z        DEBUG   consensus/consensus.go:207      received message        {"from": 3, "type": "PrepareRequest"}
2020-01-31T17:57:06.001Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "PrepareRequest", "from": 3, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:06.023Z        DEBUG   consensus/consensus.go:254      payload is already in cache     {"hash": "da64659f7956f6d18e6164241ed50f2340cb9395a6b948c26a2d1fb841a9dd38", "type": "PrepareRequest"}
2020-01-31T17:57:07.859Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/send.go:9       broadcasting message    {"type": "PrepareResponse", "height": 6015, "view": 0}
2020-01-31T17:57:07.860Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/check.go:29     check preparations      {"hasReq": true, "count": 2, "M": 3}
2020-01-31T17:57:07.860Z        DEBUG   consensus/consensus.go:207      received message        {"from": 2, "type": "RecoveryRequest"}
2020-01-31T17:57:07.860Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "RecoveryRequest", "from": 2, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:07.886Z        DEBUG   consensus/consensus.go:207      received message        {"from": 1, "type": "PrepareResponse"}
2020-01-31T17:57:07.887Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "PrepareResponse", "from": 1, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:07.887Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:389     ignoring PrepareResponse because of view changing
2020-01-31T17:57:08.135Z        DEBUG   consensus/consensus.go:207      received message        {"from": 2, "type": "PrepareResponse"}
2020-01-31T17:57:08.135Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "PrepareResponse", "from": 2, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:08.135Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:389     ignoring PrepareResponse because of view changing
2020-01-31T17:57:08.328Z        DEBUG   consensus/consensus.go:207      received message        {"from": 1, "type": "Commit"}
2020-01-31T17:57:08.328Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "Commit", "from": 1, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:08.329Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/check.go:57     not enough to commit    {"count": 1}
2020-01-31T17:57:08.461Z        DEBUG   consensus/consensus.go:207      received message        {"from": 2, "type": "Commit"}
2020-01-31T17:57:08.462Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "Commit", "from": 2, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:08.463Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/check.go:57     not enough to commit    {"count": 2}
2020-01-31T17:57:08.628Z        DEBUG   consensus/consensus.go:254      payload is already in cache     {"hash": "abcb3f58f6d1d0605ed799ec997d3372204f2aedece4179493465b6b1b6b203b", "type": "Commit"}
2020-01-31T17:57:13.020Z        DEBUG   consensus/consensus.go:207      received message        {"from": 1, "type": "RecoveryRequest"}
2020-01-31T17:57:13.020Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "RecoveryRequest", "from": 1, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:13.515Z        DEBUG   consensus/consensus.go:207      received message        {"from": 2, "type": "RecoveryRequest"}
2020-01-31T17:57:13.515Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "RecoveryRequest", "from": 2, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:13.535Z        DEBUG   consensus/consensus.go:254      payload is already in cache     {"hash": "faf148f99d040364c5261f8956f7de968023b6e87a0fba869ed7a59a5f1ee167", "type": "RecoveryRequest"}
2020-01-31T17:57:13.564Z        DEBUG   consensus/consensus.go:207      received message        {"from": 1, "type": "RecoveryMessage", "#preparation": 4, "#commit": 2, "#changeview": 0, "#request": true, "#hash": true}
2020-01-31T17:57:13.564Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:203     received message        {"type": "RecoveryMessage", "from": 1, "height": 6015, "view": 0, "my_height": 6015, "my_view": 0}
2020-01-31T17:57:13.564Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:492     recovery message received       {"dump": {"Type":65,"ViewNumber":0,"Witness":{"invocation":"40f34d455dedabb0a22fe289ce15473ee32798540a903e458313d23efd0983f51f74f51d5edf938af30a3ad03413058fa5473bed28569748085637ecdedc92fa1f","verification":"2102103a7f7dd016558597f7960d27c516a4394fd968b9e65155eb4b013e4040406eac"}}}
2020-01-31T17:57:13.564Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:389     ignoring PrepareResponse because of view changing
2020-01-31T17:57:13.564Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:393     prepare response
2020-01-31T17:57:13.568Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/check.go:29     check preparations      {"hasReq": true, "count": 3, "M": 3}
2020-01-31T17:57:13.568Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/send.go:9       broadcasting message    {"type": "Commit", "height": 6015, "view": 0}
2020-01-31T17:57:13.568Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/dbft.go:552     reset timer     {"h": 6015, "v": 0, "delay": "5s"}
2020-01-31T17:57:13.569Z        DEBUG   dbft@v0.0.0-20200131143342-b135b975faf3/check.go:64     approving block {"height": 6015, "hash": "34c4bc3c1ee70e0a88e71d5f76bc8a57f28bcad201b0188bc1c7e6756e1d865e", "tx_count": 8782, "merkle": "5537d7c49404eec163d0777dab31d036fe7076192d84851cfe835a78c9933abd", "prev": "3c480aaef10e20aba84584b27fc31eb1bcc6d421bfd3b056f567f667fda902e1"}
2020-01-31T17:57:13.664Z        DEBUG   consensus/consensus.go:254      payload is already in cache     {"hash": "018793dd37ede8896ad54da871ec68860c4adbbb559f71d42aa463710b1b0b46", "type": "RecoveryMessage"}
2020-01-31T17:57:16.047Z        INFO    blockchain persist completed    {"persistedBlocks": 1, "persistedKeys": 26348, "headerHeight": 6015, "blockHeight": 6015, "took": "61.589366ms"}
```

**Comment by fyrchik** [comment]
Fixed in #22 .
